Routing

cryogenian opened the pull request.

jdegoes commented:
Review by @garyb / @paf31. Will review myself as well.

garyb commented on bower.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "purescript-routing",
+  "version": "0.1.0",
```
Version fields in `bower.json` are redundant at the moment as it only uses git tags to resolve released versions anyway.

cryogenian replied:
bower warns if version in bower.json and tag are different.

jdegoes replied:
But if you delete it, then it won't warn at all.

cryogenian replied:
removed version

garyb commented on package.json:
```diff
@@ -0,0 +1,27 @@
+  "name": "purescript-routing",
+  "version": "0.1.0",
+  "description": "Routing lib for purescript",
+  "main": "index.js",
```
I don't think this `main` exists, does it? ;)

Usually I add a `"private": true` to `package.json` for PS projects too, as we don't distribute via npm.

cryogenian replied:
removed, added `private`

garyb commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+
+type RouteParser = P.ParserT String StateObj Checks
+type Check = Tuple String (M.StrMap String)
+type Checks = Tuple Check [Check]
```
Where does the name `Check` come from here? I think something like `RouteMatch` would be more intuitive.

cryogenian replied:
renamed

garyb commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+
+runRoute :: String -> Route -> PErr Checks
+runRoute s (Route p) =
+  case runState (P.runParserT s p) M.empty of
```
You could just use `evalState (P.runParserT s p)` here I think.

cryogenian replied:
replaced

garyb commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+    var currentHash = document.location.hash;
+    handler("")(currentHash)();
+    window.addEventListener("hashchange", function(ev) {
+      handler(ev.oldURL)(ev.newURL)();
```
`oldURL` and `newURL` aren't supported in IE8 but I take it we're not too concerned about that?

cryogenian replied:
we use file api in file component :)

garyb commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+    cb (replace rgx "" old) (replace rgx "" new)
+
+class RouteMsg a where 
+  toMsg :: Check -> Maybe a
```
Perhaps these would be better named `RouteHandler` / `handleRoute`?

Although maybe that's not quite right either, given that this is for handling a part of the route changing at a time.

cryogenian replied:
renamed to `RouteDiff` and `fromMatch`.

garyb commented on src/Routing/Parser.purs:
```diff
@@ -0,0 +1,116 @@
+
+
+-- | Ast of parsed route template
+data TemplateEl = Placeholder String | Key String | Ask [String]
```
Perhaps `Key` should be `Section` and `Ask` `Parameter` or `Query`?

cryogenian replied:
renamed to `Variable` and `Query`

garyb commented on src/Routing/Parser.purs:
```diff
@@ -0,0 +1,116 @@
+        Nothing -> do
+          lift $ modify (M.insert str v)
+          parse ts
+        _ -> P.fail $ "duplicated key '" <> str <> "' is unmatching"
```
I'm not sure I'd know what this error means without reading the code here.

garyb commented:
Sorry if it seems I'm being picky about the naming of things :) I just wasn't sure what they meant without having to dig through the code a bit.

paf31 commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+hashes :: forall e. (String -> String -> Eff e Unit) -> Eff e Unit
+hashes cb =
+  hashChanged $ \old new -> do
+    cb (replace rgx "" old) (replace rgx "" new)
```
Extract function `dropHash = replace rgx ""`?

paf31 commented:
I'll make specific comments on the individual lines, but what I'll say here is more general. This is going to seem a bit ranty, for which I apologize, but I've been planning to write something like this up for a while.

_Edit_: this has turned into a DSL tutorial, sorry. Take it for what it is - I think this would be a nice way to think about the problem.

This library seems like a really good candidate for a "layers of abstraction" approach. I'll explain what I mean by that. The pieces are all there, but I think there may be a better way to arrange them. 

- Generally, the goal is going to be to extract a pure component from your library, and a thin impure component, which is basically going to amount to calling `hashChanged`. 
- We can go further, and split the pure component into layers, each building on the last one, adding specific use cases - layers of DSLs of increasing levels of specificity, with type class laws providing the logical glue between layers. 
- The layers are modules, and module exports (including instances) define how layers interact at the boundaries, so choosing module exports is important.

In detail:

- At the bottom, we have an AST `Route` for routes. `Route`s are made up of `Path` components and `Parameter`s. You already have this in a slightly different form, and it looks fine, but you might want to rename a few things, or zip keys and values into `Tuple`s. Type classes give us general ways of constructing `Route`s:
  - `Semigroup`: `Path` is a semigroup, and so is `Route`, where we glue paths end-to-end, and concatenate parameters.
- We need to be able to parse routes from strings. `purescript-parsing` is a nice option here, but you might be able to implement something with regexes or just `String.split`.
- We need a type of errors which will be raised when a match fails. We want:
  - `Semigroup`: to combine multiple errors for a single match
  - `Semiring`: to combine errors "in parallel" - when we have two different matching functions, and they both fail.
  - We can choose a very simple implementation of errors: choose some base error type `E` and then form `type MatchError = [[E]]`. 
- Next, we need a way to match routes. What is the denotation of such a matching function? On the level of types, it is `Route -> Maybe a`, where `a` is some useful data structure that we want to get out, but we want more structure. Again, what type classes do we want?
  - `Functor`: we should be able to map over the `a`s
  - `Apply`: we should be able to pair up two consecutive matches
  - `Applicative`: we want trivial matches
  - `Alt`: we want to be able to try two alternative routes and take the first one which succeeds
  - `Alternative`: we want to be able to define filters which always fail
  - `Bind` or `Monad`? I would argue that we don't need these. `Monad` allows us to define matching functions which branch depending on a component of the route, but we can describe any finite amount of branching with `Alt`, and routes are finite, so the extra expressiveness isn't useful. Also, applicative matching functions can be statically analysed, but monadic matching functions cannot. You can't automatically derive documentation for monadic routes for example.
  - We need some "atoms" in our language which we can glue together with the instances above:
    - `lit :: String -> Match Unit` - match a literal string
    - `var :: Match String` - bind a variable
    - `param :: String -> Match String` - match a specific key
    - `anyParam :: Match (Tuple String String)` - match any key/value pair
- We can use classes to define this without choosing a representation. This is the "final" representation:

```purescript
class (Alternative f) <= Match f where
  lit :: String -> f Unit
  var :: f String
  param :: String -> f String
  anyParam :: f (Tuple String String)
```

We can already start writing matching functions without choosing a representation. Laws tell us what these routes mean:

```purescript
data FooBar = Foo String | Bar Number Boolean

route :: forall f. (Match f) => f FooBar
route = Foo <$> (lit "foo" *> var)
    <|> Bar <$> (lit "bar" *> (num <$> var)) <*> (bool <$> param "baz")
```

- Now we need to provide a model for our theory. There are many models: we can choose a model which creates Markdown documentation, or even one which works in reverse and synthesizes example routes with placeholders for variables and parameters, but the one we need is going to match routes. So choose a very naive representation to begin with, like

```purescript
newtype MatchImpl a = MatchImpl (Route -> Either MatchError a)
```

We want to be able to write `instance Match MatchImpl`. We can use instances at the lower level to define the instances we need here: `Semigroup` gives us `Applicative` and `Semiring` gives us `Alternative`.

Then we can write `runMatch :: forall a. (forall f. (Match f) => f a) -> Route -> Either MatchError a`. This is the function which we expose from our module.

- The final layer of abstraction is just cosmetic: we can write operators, type classes and functions which allow us to construct route matching functions in more idiomatic ways:

```purescript
route = ("foo" </> var) <|> ("bar" </> var <?> "baz")
```

This is probably the least important layer from the point of view of the theory, but quite important when it comes to using the library. The important thing is that we're building a language on top of an existing theory, which provides laws to guide us.

Other comments:

- Names are going to be important because there are some concepts which overlap a bit. A `Route` is a parser in your docs, for example, but there is also the concept of the "current route".

paf31 commented on src/Routing/Parser.purs:
```diff
@@ -0,0 +1,116 @@
+-- | parses variables 
+variable :: forall m. (Monad m) => P.ParserT String m String
+variable = do
+  strs <- many $ P.noneOf [":", "&", "/", "?", "="]
```
Do you need to use `decodeURIComponent` or something similar here?

cryogenian replied:
added

paf31 commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+or (Route one) (Route two) =
+  Route $ one <|> two
+
+contains :: Route -> Route -> Route
```
`Semigroup`?

cryogenian replied:
added

paf31 commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+        pure $ Tuple (Tuple name res) []
+  pure $ Route p
+
+or :: Route -> Route -> Route
```
`Semiring`?

cryogenian replied:
* `or` is commutative monoid over `Router` 
* `contains` can be monoid over `Router`, but not commutative. 
* `or` can be distributed over `contains`
* `contains` can not be distributed over `or`

`Semiring` laws says that monoid distributed over commutative monoid. 
So, it's not `Semiring` :(

paf31 replied:
Oops, I got things the wrong way round. I meant that `contains` should give `*` and `or` should give `+`. Then you have distributivity (`a*(b+c)=a*b+a*b`), and if you think of a _set_ of alternatives, and a corresponding _set_ of errors, you have commutativity too.

cryogenian replied:
ok, I'm going to add `Semiring` instance. Just need to figure out what is `zero`

paf31 replied:
Well `zero` should always fail, and `one` should always succeed, but I'm not sure what the result is.

cryogenian replied:
added

paf31 commented on src/Routing/Getter.purs:
```diff
@@ -0,0 +1,123 @@
+type Check = Tuple String (M.StrMap String)
+type Checks = Tuple Check [Check]
+type PErr a = Either P.ParseError a 
+newtype Route = Route RouteParser
```
Why the separate `newtype`?

cryogenian replied:
* to hide representation 
* to make instances

paf31 replied:
Sorry, I misread `RouteParser`.

paf31 commented on src/Routing/Parser.purs:
```diff
@@ -0,0 +1,116 @@
+
+
+-- | Produce parsers of uri from template strings
+parse :: Template -> P.ParserT String StateObj Unit
```
Why not parse the string to an intermediate representation first? i.e. separate lexing and parsing. You will have a convenient place to do things like unescaping strings and finding duplicate keys.

cryogenian replied:
actually it's lexer. Parser is `template` function

paf31 replied:
Well they're both `ParserT String _ _`. If there is a lexer there would be an intermediate data structure.

cryogenian replied:
Heh, maybe it's interpreter then, but it's not parser definitely, since it produces parsers, that will parse url

jdegoes commented:
Awesome feedback, @paf31. Maxim, think you can make these changes or would you like some assistance / guidance from @paf31?

cryogenian commented:
* renamed some function, hope their names are better now
* remove unneeded fields from `package.json`
* added `decodeURIComponent` 
* added monoid to `Router`

cryogenian commented:
@paf31 
* `purescript-parsing` handles not only template parsing but error composition too.
* AST of router is linear, so, there is no need to handle error parallel.
* `ParserT` already has all this behaviours (`Alt`, `Alternative`, `Functor`, etc) 
* Signals from `hashchange` can be invariant with application state. But I believe that they not have to. Especially in cases of handling subroutes, where one can want (this one is me, though) not to emit messages of parent state. 
* Adding types to those messages (like `bool` or `num`) can be made by adding specific instances for `RouteDiff` now. I suggest that adding specific parser for this case will work also.

paf31 commented:
> AST of router is linear, so, there is no need to handle error parallel.

In a particular template, yes, but what if you want to handle two different types of routes? A `Semiring` of errors will handle distributing the sequential errors over the alternatives so that you'll end up with something like

```
(Expected "/foo") + (Expected "/bar" * RequiredParam "baz")
```

which you would then be able to postprocess to a more meaningful error message like:

```
Expected one of "/foo" or "/bar"
```

cryogenian commented:
@jdegoes 
I hope I can make this changes myself. Or maybe with little help of @paf31. And make them if you want.

But, personally, I don't see the point:
* Lib can handle routes 
* Lib can handle subroutes
* Lib emits typesafe messages (more or less, though, if one don't provide `RouteDiff` instances there will be only `Tuple String (StrMap String)` 
* Lib handles setting of specific states as hash.

cryogenian commented:
@paf31 
Why do you need this messages? I think it's not very difficult to make them in my representation by adding field to `data Router = Router {error :: String, parser :: ParserT ...}`

paf31 commented:
Yeah, as I say, just understand my notes as an alternative approach, and to a certain extent, me trying to understand the problem. If everything works as needed, maybe don't worry about it. I put the more important items (`decode`, etc.) that I saw in line comments.

cryogenian commented:
@paf31 Thank you very much for review! It's very helpful.

jdegoes commented:
> But, personally, I don't see the point:

It's a free lesson in how to organize a library from the creator of PureScript. :smiley: 

I think @paf31's suggestions would improve the organization of the library even if they wouldn't add functionality. If you agree, then go ahead and make them, if you don't agree, well we can discuss further on Skype or just merge the PR as-is.

cryogenian commented:
Great! I'll try to try @paf31 approach in different branch and make pr in day or two
